### PR TITLE
gh-92877: Advise to use typing.IO sparingly

### PR DESCRIPTION
**Pull request by srittau**

### CODE REVIEW

File: Doc/library/typing.rst [hunk 1]
```diff
@@ -1940,6 +1940,25 @@ Other concrete types
...
 
+   These types should be used sparingly. For argument types, a tight
+   :class:`Protocol` that documents the actual needs of the implementation
+   is often a better alternative. Return values should be annotated with
+   the actual concrete type returned. For example::
+
+       from io import StringIO
+
+       class Reader(Protocol):
+           def read(self, __size: int) -> bytes: ...
```
Comment (JelleZijlstra) [inline Doc/library/typing.rst]:
```suggestion
           def read(self, size: int, /) -> bytes: ...
```

Comment (srittau) [reply Doc/library/typing.rst]: Oh, how I wish we could use that in stubs already. :)

**Review by JelleZijlstra — COMMENTED** [review]
While using Protocols works well for typeshed, I'm not sure it's the right advice for all typing users. Protocols are an advanced feature and the details are often tricky to get right. It's mostly fine for users to use `IO`/`TextIO`/`BinaryIO` within applications where they may not need to worry about exotic file-like objectts.

A friendlier approach could be along the lines of "Not all file-like types are recognized as instances of these types. If you encounter such types, consider using a Protocol instead."

**Comment by srittau** [comment]
> While using Protocols works well for typeshed, I'm not sure it's the right advice for all typing users. Protocols are an advanced feature and the details are often tricky to get right. It's mostly fine for users to use `IO`/`TextIO`/`BinaryIO` within applications where they may not need to worry about exotic file-like objectts.
> 
> A friendlier approach could be along the lines of "Not all file-like types are recognized as instances of these types. If you encounter such types, consider using a Protocol instead."

Maybe it would be a good idea to provide slimmed down protocols, like those suggested in python/typing#564. I think we by now have a good sense for what methods are commonly accessed. We could then suggest to use those protocols.

**Comment by srittau** [comment]
I'm closing this for now while exploring other approaches.

**Pull request closed without merging**